File: src/ssh_manager/core/manifest.py
"""Manifest domain models. Validated with pydantic v2.

The manifest is the single source of truth (invariant 1). These models load/save
through the atomic JSON store and expose the per-host key resolution the renderer
and reconciler depend on (``per_service`` default, ``shared`` opt-in).
"""
from __future__ import annotations

from collections.abc import Iterator
from pathlib import Path
from typing import Any

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from ..util import jsonstore
from ..util.errors import ManifestError
from .key import derive_key_name

SCHEMA_VERSION = 1
SSH_TOKEN = "~/.ssh"  # IdentityFile paths in rendered config use the ~ form

# --- input hardening -------------------------------------------------------
# The manifest is rendered verbatim into ~/.ssh/config and its names become
# filesystem paths, so untrusted values (a hand-edited manifest, an imported
# foreign ssh config) must not inject config directives, escape the key tree, or
# smuggle leading-dash arguments into ssh.
import re as _re  # noqa: E402

_CONTROL = _re.compile(r"[\x00-\x1f\x7f]")          # newline/tab/NUL/other control

# ssh options that run a command, load a shared object, or pull in further config
# - never allowed in raw_options / global_options. ProxyJump (a host, not a
# command) stays allowed. This is a denylist of the known code/config-execution
# directives; values are additionally guarded for control chars + leading dash.
_DANGEROUS_OPTIONS = frozenset({
    "proxycommand", "localcommand", "permitlocalcommand", "remotecommand",
    "match", "include", "knownhostscommand", "pkcs11provider", "securitykeyprovider",
})


def _reject_control(field: str, value: str) -> str:
    if _CONTROL.search(value):
        raise ValueError(f"{field} contains a control character or newline")
    return value


def _safe_segment(field: str, value: str) -> str:
    """A safe single filesystem path component + ssh token: no traversal, no
    separators, no leading dash, no control chars, no whitespace (would split into
    two config tokens / two key paths), and no glob metacharacters ``*?`` (an
    alias of ``*`` would render a wildcard ``Host`` block that overrides others)."""
    _reject_control(field, value)
    if (value in ("", ".", "..") or "/" in value or "\\" in value
            or value.startswith("-") or any(c.isspace() for c in value)
            or any(c in value for c in "*?")):
        raise ValueError(
            f"{field}={value!r} is not a safe name "
            "(no empty/'.'/'..'/'/'/'\\\\'/leading '-'/whitespace/'*'/'?')")
    return value


def _safe_value(field: str, value: str) -> str:
    """A non-path string still rendered into config / passed to ssh: no control
    chars, no leading dash (would be parsed as an ssh option), no whitespace (a
    hostname/user with a space would split into extra config tokens)."""
    _reject_control(field, value)
    if value.startswith("-"):
        raise ValueError(f"{field}={value!r} must not start with '-'")
    if any(c.isspace() for c in value):
        raise ValueError(f"{field}={value!r} must not contain whitespace")
    return value


_KEY_SCOPES = frozenset({"per_service", "shared"})


def _check_key_scope(value: str) -> str:
    if value not in _KEY_SCOPES:
        raise ValueError(f"key_scope must be one of {sorted(_KEY_SCOPES)} (got {value!r})")
    return value


def _check_options(field: str, opts: dict[str, str]) -> dict[str, str]:
    for k, v in opts.items():
        _reject_control(f"{field} key {k!r}", k)
        _reject_control(f"{field}[{k}]", v)
        if k.lower() in _DANGEROUS_OPTIONS:
            raise ValueError(f"{field} key {k!r} is not allowed (it can execute commands)")
    return opts

# Canonical Host* defaults. UseKeychain is dropped off macOS by the
# renderer; IgnoreUnknown must precede the option it guards.
DEFAULT_GLOBAL_OPTIONS: dict[str, str] = {
    "AddKeysToAgent": "yes",
    "IgnoreUnknown": "UseKeychain",
    "UseKeychain": "yes",
    "IdentitiesOnly": "yes",
    "ServerAliveInterval": "60",
}


class Host(BaseModel):
    model_config = ConfigDict(extra="forbid")

    alias: str
    hostname: str
    user: str
    port: int = 22
    provider: str | None = None          # named adapter; else generic SSH
    token_env: str | None = None         # per-host provider credential ref
    key_name: str | None = None          # None when profile key_scope == "shared"
    tags: list[str] = Field(default_factory=list)
    requires_vpn: bool = False           # host is only reachable over a VPN
    vpn_name: str | None = None          # which VPN (shown in the reachability hint)
    vpn_url: str | None = None           # where to connect that VPN (shown in the hint)
    raw_options: dict[str, str] = Field(default_factory=dict)

    @field_validator("raw_options", mode="before")
    @classmethod
    def _stringify_raw(cls, value: Any) -> Any:
        if isinstance(value, dict):
            return {k: str(v) for k, v in value.items()}
        return value

    @field_validator("alias", "key_name")
    @classmethod
    def _v_segment(cls, value: str | None, info: Any) -> str | None:
        return value if value is None else _safe_segment(info.field_name, value)

    @field_validator("hostname", "user")
    @classmethod
    def _v_value(cls, value: str, info: Any) -> str:
        return _safe_value(info.field_name, value)

    @field_validator("raw_options")
    @classmethod
    def _v_raw_options(cls, value: dict[str, str]) -> dict[str, str]:
        return _check_options("raw_options", value)


class Profile(BaseModel):
    model_config = ConfigDict(extra="forbid")

    key_scope: str = "per_service"          # "per_service" | "shared"
    key_name: str | None = None          # used when key_scope == "shared"
    hosts: list[Host] = Field(default_factory=list)

    @field_validator("key_name")
    @classmethod
    def _v_key_name(cls, value: str | None) -> str | None:
        return value if value is None else _safe_segment("profile key_name", value)

    @field_validator("key_scope")
    @classmethod
    def _v_key_scope(cls, value: str) -> str:
        return _check_key_scope(value)


class ExpiryCheck(BaseModel):
    model_config = ConfigDict(extra="forbid")

    enabled: bool = True
    debounce_hours: int = 24
    desktop_notify: bool = True


class Defaults(BaseModel):
    model_config = ConfigDict(extra="forbid")

    key_type: str = "ed25519"
    key_scope: str = "per_service"
    rotate_after_days: int = 365
    warn_before_days: list[int] = Field(default_factory=lambda: [30, 14, 7, 1])
    expiry_check: ExpiryCheck = Field(default_factory=ExpiryCheck)
    global_options: dict[str, str] = Field(default_factory=dict)

    @field_validator("global_options", mode="before")
    @classmethod
    def _stringify_options(cls, value: Any) -> Any:
        """ssh option values may be numbers in JSON (e.g. ServerAliveInterval: 60);
        render needs strings, so coerce here at the edge."""
        if isinstance(value, dict):
            return {k: str(v) for k, v in value.items()}
        return value

    @field_validator("global_options")
    @classmethod
    def _v_global_options(cls, value: dict[str, str]) -> dict[str, str]:
        return _check_options("global_options", value)

    @field_validator("key_scope")
    @classmethod
    def _v_key_scope(cls, value: str) -> str:
        return _check_key_scope(value)


class ResolvedKey(BaseModel):
    """A host paired with its resolved key name + IdentityFile path."""

    profile: str
    host: Host
    key_name: str
    identity_file: str          # e.g. ~/.ssh/profiles/work/work_unc-ed25519


class Manifest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    version: int = SCHEMA_VERSION
    defaults: Defaults = Field(default_factory=Defaults)
    profiles: dict[str, Profile] = Field(default_factory=dict)

    @field_validator("profiles")
    @classmethod
    def _v_profile_names(cls, value: dict[str, Profile]) -> dict[str, Profile]:
        for name in value:                       # profile names become directory paths
            _safe_segment("profile name", name)
        return value

    @model_validator(mode="after")
    def _v_key_name_uniqueness(self) -> Manifest:
        """A key_name must belong to exactly one profile. rotate/deploy/rollback
        resolve a key_name back to its host(s) and assume they all share ONE
        profile dir; if two profiles reused a name (e.g. two ``shared`` profiles
        both named their key ``deploy``), rotating one would mint into one profile's
        dir yet deploy to the other profile's hosts - orphaning/locking them out.
        Reject that at load so the invariant the lifecycle ops rely on holds."""
        owner: dict[str, str] = {}
        for pname, profile in self.profiles.items():
            for host in profile.hosts:
                try:
                    kname = self.resolved_key_name(pname, host)
                except ManifestError:
                    # An unresolvable key (e.g. shared profile missing key_name) is
                    # reported by resolved_key_name at use-time; don't pre-empt that
                    # distinct, more specific error here.
                    continue
                prev = owner.setdefault(kname, pname)
                if prev != pname:
                    raise ManifestError(
                        f"key_name {kname!r} is used by both profile {prev!r} and "
                        f"{pname!r}; a key_name must be unique across profiles "
                        "(rename one, or namespace it per profile)")
        return self

    # Repository pattern (atomic, version-stamped)
    @classmethod
    def load(cls, path: str | Path) -> Manifest:
        path = Path(path)
        if not path.exists():
            raise ManifestError(f"manifest not found: {path} (run: sshmgr init)")
        try:
            data = jsonstore.read_json(path)
        except ValueError as exc:
            raise ManifestError(f"manifest is not valid JSON: {path}: {exc}") from exc
        except OSError as exc:
            raise ManifestError(f"manifest could not be read: {path}: {exc}") from exc
        try:
            return cls.model_validate(data)
        except Exception as exc:  # pydantic ValidationError -> our typed error
            raise ManifestError(f"manifest failed validation: {exc}") from exc

    def save(self, path: str | Path) -> None:
        jsonstore.write_json_atomic(path, self.model_dump(mode="json"))

    @classmethod
    def starter(cls, *, emit_use_keychain: bool = True) -> Manifest:
        """A minimal valid manifest for `sshmgr init` - defaults + no profiles."""
        opts = dict(DEFAULT_GLOBAL_OPTIONS)
        if not emit_use_keychain:
            opts.pop("UseKeychain", None)
        return cls(defaults=Defaults(global_options=opts), profiles={})

    # key resolution (per_service default, shared opt-in)
    def resolved_key_name(self, profile_name: str, host: Host) -> str:
        profile = self.profiles[profile_name]
        if profile.key_scope == "shared":
            if not profile.key_name:
                raise ManifestError(
                    f"profile {profile_name!r} is shared but sets no key_name"
                )
            return profile.key_name
        if host.key_name:
            return host.key_name
        algo = self.defaults.key_type
        return derive_key_name(profile_name, host.alias, algo)

    def identity_file(self, profile_name: str, key_name: str) -> str:
        return f"{SSH_TOKEN}/profiles/{profile_name}/{key_name}"

    def known_hosts_file(self, profile_name: str) -> str:
        """Per-profile host-key trust store - everything lives under the profile,
        so a host key trusted in one identity context never bleeds into another."""
        return f"{SSH_TOKEN}/profiles/{profile_name}/known_hosts"

    def iter_resolved(self) -> Iterator[ResolvedKey]:
        """Yield every host with its resolved key, skipping empty profiles."""
        for pname, profile in self.profiles.items():
            for host in profile.hosts:
                kname = self.resolved_key_name(pname, host)
                yield ResolvedKey(
                    profile=pname, host=host, key_name=kname,
                    identity_file=self.identity_file(pname, kname),
                )

    def non_empty_profiles(self) -> list[str]:
        return [name for name, p in self.profiles.items() if p.hosts]

File: internal/core/manifest/manifest.go
// Package manifest is the ssh-manager manifest domain model, ported from
// src/ssh_manager/core/manifest.py. The manifest is the single source of truth;
// this package loads/validates it and exposes the per-host key resolution the
// renderer and reconciler depend on (per_service default, shared opt-in).
package manifest

import (
	"bytes"
	"encoding/json"
	"fmt"
	"os"
	"regexp"
	"sort"
	"strings"
	"unicode"

	"github.com/simtabi/ssh-manager/internal/core/key"
)

const (
	schemaVersion = 1
	sshToken      = "~/.ssh" // IdentityFile paths render in the ~ form
)

var controlChars = regexp.MustCompile(`[\x00-\x1f\x7f]`)

// dangerousOptions run a command, load an object, or pull in further config and
// are never allowed in raw_options / global_options (ProxyJump is a host, allowed).
var dangerousOptions = map[string]bool{
	"proxycommand": true, "localcommand": true, "permitlocalcommand": true,
	"remotecommand": true, "match": true, "include": true,
	"knownhostscommand": true, "pkcs11provider": true, "securitykeyprovider": true,
}

var keyScopes = map[string]bool{"per_service": true, "shared": true}

// DefaultGlobalOptions are the canonical Host* defaults used by Starter.
var DefaultGlobalOptions = map[string]string{
	"AddKeysToAgent": "yes", "IgnoreUnknown": "UseKeychain", "UseKeychain": "yes",
	"IdentitiesOnly": "yes", "ServerAliveInterval": "60",
}

// OrderedOptions is an SSH-option map that preserves JSON key order (the renderer
// emits options in that order) and stringifies values like Python's str().
type OrderedOptions struct {
	keys []string
	vals map[string]string
}

func (o *OrderedOptions) UnmarshalJSON(b []byte) error {
	o.keys = nil
	o.vals = map[string]string{}
	dec := json.NewDecoder(bytes.NewReader(b))
	t, err := dec.Token()
	if err != nil {
		return err
	}
	if t == nil {
		return nil // null -> empty
	}
	if d, ok := t.(json.Delim); !ok || d != '{' {
		return fmt.Errorf("options must be a JSON object")
	}
	for dec.More() {
		kt, err := dec.Token()
		if err != nil {
			return err
		}
		k := kt.(string)
		var raw json.RawMessage
		if err := dec.Decode(&raw); err != nil {
			return err
		}
		if _, seen := o.vals[k]; !seen {
			o.keys = append(o.keys, k)
		}
		o.vals[k] = stringifyJSON(raw)
	}
	_, err = dec.Token() // consume '}'
	return err
}

// MarshalJSON emits the options in their preserved order.
func (o OrderedOptions) MarshalJSON() ([]byte, error) {
	if len(o.keys) == 0 {
		return []byte("{}"), nil
	}
	var buf bytes.Buffer
	buf.WriteByte('{')
	for i, k := range o.keys {
		if i > 0 {
			buf.WriteByte(',')
		}
		kb, _ := json.Marshal(k)
		vb, _ := json.Marshal(o.vals[k])
		buf.Write(kb)
		buf.WriteByte(':')
		buf.Write(vb)
	}
	buf.WriteByte('}')
	return buf.Bytes(), nil
}

// Len, Keys, and Get expose the options in order.
func (o OrderedOptions) Len() int            { return len(o.keys) }
func (o OrderedOptions) Keys() []string      { return o.keys }
func (o OrderedOptions) Get(k string) string { return o.vals[k] }

func stringifyJSON(raw json.RawMessage) string {
	s := strings.TrimSpace(string(raw))
	if len(s) >= 1 && s[0] == '"' {
		var str string
		_ = json.Unmarshal(raw, &str)
		return str
	}
	switch s {
	case "true":
		return "True" // match Python str(True)
	case "false":
		return "False"
	case "null":
		return "None"
	}
	return s // number token, e.g. "60"
}

// Host is a single SSH host entry.
type Host struct {
	Alias       string         `json:"alias"`
	Hostname    string         `json:"hostname"`
	User        string         `json:"user"`
	Port        int            `json:"port"`
	Provider    *string        `json:"provider,omitempty"`
	TokenEnv    *string        `json:"token_env,omitempty"`
	KeyName     *string        `json:"key_name,omitempty"`
	Tags        []string       `json:"tags,omitempty"`
	RequiresVPN bool           `json:"requires_vpn"`
	VPNName     *string        `json:"vpn_name,omitempty"`
	VPNURL      *string        `json:"vpn_url,omitempty"`
	RawOptions  OrderedOptions `json:"raw_options,omitempty"`
}

func (h *Host) UnmarshalJSON(b []byte) error {
	type alias Host
	aux := alias{Port: 22}
	if err := decodeStrict(b, &aux); err != nil {
		return err
	}
	*h = Host(aux)
	return nil
}

// MarshalJSON emits every field in declaration order (null for unset pointers, []
// for no tags), matching pydantic model_dump(mode="json"). raw_options serializes
// {} when empty via OrderedOptions.
func (h Host) MarshalJSON() ([]byte, error) {
	type wire struct {
		Alias       string         `json:"alias"`
		Hostname    string         `json:"hostname"`
		User        string         `json:"user"`
		Port        int            `json:"port"`
		Provider    *string        `json:"provider"`
		TokenEnv    *string        `json:"token_env"`
		KeyName     *string        `json:"key_name"`
		Tags        []string       `json:"tags"`
		RequiresVPN bool           `json:"requires_vpn"`
		VPNName     *string        `json:"vpn_name"`
		VPNURL      *string        `json:"vpn_url"`
		RawOptions  OrderedOptions `json:"raw_options"`
	}
	tags := h.Tags
	if tags == nil {
		tags = []string{}
	}
	return json.Marshal(wire{
		Alias: h.Alias, Hostname: h.Hostname, User: h.User, Port: h.Port,
		Provider: h.Provider, TokenEnv: h.TokenEnv, KeyName: h.KeyName, Tags: tags,
		RequiresVPN: h.RequiresVPN, VPNName: h.VPNName, VPNURL: h.VPNURL, RawOptions: h.RawOptions,
	})
}

// Profile groups hosts that share an identity.
type Profile struct {
	KeyScope string  `json:"key_scope"`
	KeyName  *string `json:"key_name,omitempty"`
	Hosts    []Host  `json:"hosts"`
}

func (p *Profile) UnmarshalJSON(b []byte) error {
	type alias Profile
	aux := alias{KeyScope: "per_service"}
	if err := decodeStrict(b, &aux); err != nil {
		return err
	}
	*p = Profile(aux)
	return nil
}

// MarshalJSON emits key_scope, key_name (null when unset), and hosts ([] when
// none) in declaration order, matching pydantic.
func (p Profile) MarshalJSON() ([]byte, error) {
	type wire struct {
		KeyScope string  `json:"key_scope"`
		KeyName  *string `json:"key_name"`
		Hosts    []Host  `json:"hosts"`
	}
	hosts := p.Hosts
	if hosts == nil {
		hosts = []Host{}
	}
	return json.Marshal(wire{KeyScope: p.KeyScope, KeyName: p.KeyName, Hosts: hosts})
}

// ExpiryCheck is the notifier policy.
type ExpiryCheck struct {
	Enabled       bool `json:"enabled"`
	DebounceHours int  `json:"debounce_hours"`
	DesktopNotify bool `json:"desktop_notify"`
}

func newExpiryCheck() ExpiryCheck {
	return ExpiryCheck{Enabled: true, DebounceHours: 24, DesktopNotify: true}
}

func (e *ExpiryCheck) UnmarshalJSON(b []byte) error {
	type alias ExpiryCheck
	aux := alias(newExpiryCheck())
	if err := decodeStrict(b, &aux); err != nil {
		return err
	}
	*e = ExpiryCheck(aux)
	return nil
}

// Defaults are manifest-wide defaults.
type Defaults struct {
	KeyType         string         `json:"key_type"`
	KeyScope        string         `json:"key_scope"`
	RotateAfterDays int            `json:"rotate_after_days"`
	WarnBeforeDays  []int          `json:"warn_before_days"`
	ExpiryCheck     ExpiryCheck    `json:"expiry_check"`
	GlobalOptions   OrderedOptions `json:"global_options"`
}

func newDefaults() Defaults {
	return Defaults{
		KeyType: "ed25519", KeyScope: "per_service", RotateAfterDays: 365,
		WarnBeforeDays: []int{30, 14, 7, 1}, ExpiryCheck: newExpiryCheck(),
		GlobalOptions: OrderedOptions{},
	}
}

func (d *Defaults) UnmarshalJSON(b []byte) error {
	type alias Defaults
	aux := alias(newDefaults())
	if err := decodeStrict(b, &aux); err != nil {
		return err
	}
	*d = Defaults(aux)
	return nil
}

// MarshalJSON emits warn_before_days as [] when empty (not null) and global_options
// as {} via OrderedOptions, in declaration order.
func (d Defaults) MarshalJSON() ([]byte, error) {
	type wire struct {
		KeyType         string         `json:"key_type"`
		KeyScope        string         `json:"key_scope"`
		RotateAfterDays int            `json:"rotate_after_days"`
		WarnBeforeDays  []int          `json:"warn_before_days"`
		ExpiryCheck     ExpiryCheck    `json:"expiry_check"`
		GlobalOptions   OrderedOptions `json:"global_options"`
	}
	warn := d.WarnBeforeDays
	if warn == nil {
		warn = []int{}
	}
	return json.Marshal(wire{
		KeyType: d.KeyType, KeyScope: d.KeyScope, RotateAfterDays: d.RotateAfterDays,
		WarnBeforeDays: warn, ExpiryCheck: d.ExpiryCheck, GlobalOptions: d.GlobalOptions,
	})
}

// ResolvedKey pairs a host with its resolved key name + IdentityFile path.
type ResolvedKey struct {
	Profile      string
	Host         Host
	KeyName      string
	IdentityFile string
}

// Manifest is the whole manifest.
type Manifest struct {
	Version  int                `json:"version"`
	Defaults Defaults           `json:"defaults"`
	Profiles map[string]Profile `json:"profiles"`

	// profileOrder is the JSON insertion order of profile keys. The renderer
	// emits in name order, but the read views (query.groups) iterate the manifest
	// in file order, so we preserve it to match v1's list/view output order.
	profileOrder []string
}

func (m *Manifest) UnmarshalJSON(b []byte) error {
	type alias Manifest
	aux := alias{Version: schemaVersion, Defaults: newDefaults(), Profiles: map[string]Profile{}}
	if err := decodeStrict(b, &aux); err != nil {
		return err
	}
	*m = Manifest(aux)
	var top map[string]json.RawMessage
	if json.Unmarshal(b, &top) == nil {
		if pr, ok := top["profiles"]; ok {
			m.profileOrder = objectKeyOrder(pr)
		}
	}
	return nil
}

// MarshalJSON emits {version, defaults, profiles} with profiles in manifest (file)
// order - a Go map would otherwise marshal its keys sorted. Compact output;
// Save re-indents it to match pydantic's json.dump(indent=2).
func (m Manifest) MarshalJSON() ([]byte, error) {
	var b bytes.Buffer
	b.WriteString(`{"version":`)
	vb, err := json.Marshal(m.Version)
	if err != nil {
		return nil, err
	}
	b.Write(vb)
	b.WriteString(`,"defaults":`)
	db, err := json.Marshal(m.Defaults)
	if err != nil {
		return nil, err
	}
	b.Write(db)
	b.WriteString(`,"profiles":{`)
	for i, name := range m.ProfileNames() {
		if i > 0 {
			b.WriteByte(',')
		}
		kb, _ := json.Marshal(name)
		b.Write(kb)
		b.WriteByte(':')
		pb, err := json.Marshal(m.Profiles[name])
		if err != nil {
			return nil, err
		}
		b.Write(pb)
	}
	b.WriteString("}}")
	return b.Bytes(), nil
}

// objectKeyOrder returns a JSON object's keys in their textual order (nil if raw
// is not an object), the same token-stream technique OrderedOptions uses.
func objectKeyOrder(raw json.RawMessage) []string {
	dec := json.NewDecoder(bytes.NewReader(raw))
	t, err := dec.Token()
	if err != nil {
		return nil
	}
	if d, ok := t.(json.Delim); !ok || d != '{' {
		return nil
	}
	var keys []string
	for dec.More() {
		kt, err := dec.Token()
		if err != nil {
			return keys
		}
		keys = append(keys, kt.(string))
		var skip json.RawMessage
		if dec.Decode(&skip) != nil {
			return keys
		}
	}
	return keys
}

// ProfileNames returns profile names in manifest (file) order. Falls back to
// name order for a programmatically built manifest with no captured order.
func (m *Manifest) ProfileNames() []string {
	if len(m.profileOrder) == len(m.Profiles) {
		return m.profileOrder
	}
	return m.sortedProfileNames()
}

// SetProfile adds or replaces a profile, preserving file order (a new profile is
// appended, like Python's dict insertion).
func (m *Manifest) SetProfile(name string, p Profile) {
	if _, ok := m.Profiles[name]; !ok {
		m.profileOrder = append(m.profileOrder, name)
	}
	if m.Profiles == nil {
		m.Profiles = map[string]Profile{}
	}
	m.Profiles[name] = p
}

// DeleteProfile removes a profile and its order entry.
func (m *Manifest) DeleteProfile(name string) {
	delete(m.Profiles, name)
	for i, n := range m.profileOrder {
		if n == name {
			m.profileOrder = append(m.profileOrder[:i], m.profileOrder[i+1:]...)
			break
		}
	}
}

// Validate re-runs the manifest validators (so a bad edit can't be persisted).
func (m *Manifest) Validate() error { return m.validate() }

// decodeStrict decodes with DisallowUnknownFields (pydantic extra="forbid").
func decodeStrict(b []byte, v any) error {
	dec := json.NewDecoder(bytes.NewReader(b))
	dec.DisallowUnknownFields()
	return dec.Decode(v)
}

// --- validation (mirrors the pydantic validators) --------------------------

func rejectControl(field, value string) error {
	if controlChars.MatchString(value) {
		return fmt.Errorf("%s contains a control character or newline", field)
	}
	return nil
}

func safeSegment(field, value string) error {
	if err := rejectControl(field, value); err != nil {
		return err
	}
	bad := value == "" || value == "." || value == ".." ||
		strings.ContainsAny(value, `/\*?`) || strings.HasPrefix(value, "-")
	if !bad {
		for _, r := range value {
			if unicode.IsSpace(r) {
				bad = true
				break
			}
		}
	}
	if bad {
		return fmt.Errorf("%s=%q is not a safe name "+
			"(no empty/'.'/'..'/'/'/'\\'/leading '-'/whitespace/'*'/'?')", field, value)
	}
	return nil
}

func safeValue(field, value string) error {
	if err := rejectControl(field, value); err != nil {
		return err
	}
	if strings.HasPrefix(value, "-") {
		return fmt.Errorf("%s=%q must not start with '-'", field, value)
	}
	for _, r := range value {
		if unicode.IsSpace(r) {
			return fmt.Errorf("%s=%q must not contain whitespace", field, value)
		}
	}
	return nil
}

func checkOptions(field string, opts OrderedOptions) error {
	for _, k := range opts.keys {
		v := opts.vals[k]
		if err := rejectControl(fmt.Sprintf("%s key %q", field, k), k); err != nil {
			return err
		}
		if err := rejectControl(fmt.Sprintf("%s[%s]", field, k), v); err != nil {
			return err
		}
		if dangerousOptions[strings.ToLower(k)] {
			return fmt.Errorf("%s key %q is not allowed (it can execute commands)", field, k)
		}
	}
	return nil
}

func checkKeyScope(value string) error {
	if !keyScopes[value] {
		return fmt.Errorf("key_scope must be one of [per_service shared] (got %q)", value)
	}
	return nil
}

func (m *Manifest) validate() error {
	if err := checkOptions("global_options", m.Defaults.GlobalOptions); err != nil {
		return err
	}
	if err := checkKeyScope(m.Defaults.KeyScope); err != nil {
		return err
	}
	for name, p := range m.Profiles {
		if err := safeSegment("profile name", name); err != nil {
			return err
		}
		if err := checkKeyScope(p.KeyScope); err != nil {
			return err
		}
		if p.KeyName != nil {
			if err := safeSegment("profile key_name", *p.KeyName); err != nil {
				return err
			}
		}
		for _, h := range p.Hosts {
			if err := safeSegment("alias", h.Alias); err != nil {
				return err
			}
			if h.KeyName != nil {
				if err := safeSegment("key_name", *h.KeyName); err != nil {
					return err
				}
			}
			if err := safeValue("hostname", h.Hostname); err != nil {
				return err
			}
			if err := safeValue("user", h.User); err != nil {
				return err
			}
			if err := checkOptions("raw_options", h.RawOptions); err != nil {
				return err
			}
		}
	}
	return m.validateKeyNameUniqueness()
}

func (m *Manifest) validateKeyNameUniqueness() error {
	owner := map[string]string{}
	for _, pname := range m.sortedProfileNames() {
		for _, h := range m.Profiles[pname].Hosts {
			kname, err := m.ResolvedKeyName(pname, h)
			if err != nil {
				continue // unresolvable key reported at use-time
			}
			if prev, ok := owner[kname]; ok && prev != pname {
				return fmt.Errorf("key_name %q is used by both profile %q and %q; "+
					"a key_name must be unique across profiles", kname, prev, pname)
			}
			owner[kname] = pname
		}
	}
	return nil
}

func (m *Manifest) sortedProfileNames() []string {
	names := make([]string, 0, len(m.Profiles))
	for n := range m.Profiles {
		names = append(names, n)
	}
	sort.Strings(names)
	return names
}

// --- key resolution --------------------------------------------------------

// ResolvedKeyName resolves a host's key name (per_service derives it; shared uses
// the profile key_name).
func (m *Manifest) ResolvedKeyName(profileName string, host Host) (string, error) {
	profile, ok := m.Profiles[profileName]
	if !ok {
		return "", fmt.Errorf("no such profile: %q", profileName)
	}
	if profile.KeyScope == "shared" {
		if profile.KeyName == nil || *profile.KeyName == "" {
			return "", fmt.Errorf("profile %q is shared but sets no key_name", profileName)
		}
		return *profile.KeyName, nil
	}
	if host.KeyName != nil && *host.KeyName != "" {
		return *host.KeyName, nil
	}
	return key.DeriveKeyName(profileName, host.Alias, m.Defaults.KeyType)
}

// IdentityFile is the rendered ~ form path for a key (always forward slashes).
func (m *Manifest) IdentityFile(profileName, keyName string) string {
	return sshToken + "/profiles/" + profileName + "/" + keyName
}

// KnownHostsFile is the per-profile host-key trust store path.
func (m *Manifest) KnownHostsFile(profileName string) string {
	return sshToken + "/profiles/" + profileName + "/known_hosts"
}

// IterResolved returns every host with its resolved key, in manifest (file)
// order - matching Python's iter_resolved (which iterates profiles.items()). Order
// is observable in order-preserving consumers like doctor's unpinned-host list;
// the rendered config is order-independent (the root config uses Include).
func (m *Manifest) IterResolved() ([]ResolvedKey, error) {
	var out []ResolvedKey
	for _, pname := range m.ProfileNames() {
		for _, h := range m.Profiles[pname].Hosts {
			kname, err := m.ResolvedKeyName(pname, h)
			if err != nil {
				return nil, err
			}
			out = append(out, ResolvedKey{
				Profile: pname, Host: h, KeyName: kname,
				IdentityFile: m.IdentityFile(pname, kname),
			})
		}
	}
	return out, nil
}

// NonEmptyProfiles lists profiles that have at least one host, in manifest (file)
// order - matching Python's non_empty_profiles (profiles.items()).
func (m *Manifest) NonEmptyProfiles() []string {
	var out []string
	for _, n := range m.ProfileNames() {
		if len(m.Profiles[n].Hosts) > 0 {
			out = append(out, n)
		}
	}
	return out
}

// --- repository ------------------------------------------------------------

// Load reads and validates a manifest from path.
func Load(path string) (*Manifest, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, fmt.Errorf("manifest not found: %s (run: sshmgr init)", path)
		}
		return nil, fmt.Errorf("manifest could not be read: %s: %w", path, err)
	}
	var m Manifest
	if err := json.Unmarshal(data, &m); err != nil {
		return nil, fmt.Errorf("manifest failed validation: %s: %w", path, err)
	}
	if err := m.validate(); err != nil {
		return nil, fmt.Errorf("manifest failed validation: %w", err)
	}
	return &m, nil
}

// Save writes the manifest as indented JSON.
func (m *Manifest) Save(path string) error {
	b, err := json.MarshalIndent(m, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(path, append(b, '\n'), 0o600)
}
